source/opt/loop_unroller.cpp

```diff
@@ -420,36 +491,58 @@ void LoopUnrollerUtilsImpl::PartiallyUnrollResidualFactor(ir::Loop* loop,
       *context_->GetLoopDescriptor(&function_);
 
   loop_descriptor.AddLoop(new_loop, loop->GetParent());
+
+  RemoveDeadInstructions();
 }
 
-// Duplicate the |loop| body |factor| number of times while keeping the loop
-// backedge intact.
-void LoopUnrollerUtilsImpl::PartiallyUnroll(ir::Loop* loop, size_t factor) {
-  Unroll(loop, factor);
-  AddBlocksToLoop(loop);
-  AddBlocksToFunction(loop->GetMergeBlock());
+// Mark this loop as DontUnroll as it will already be unrolled and it may not
+// be safe to unroll a previously partially unrolled loop.
+void LoopUnrollerUtilsImpl::MarkLoopControlAsDontUnroll(ir::Loop* loop) const {
+  ir::Instruction* loop_merge_inst = loop->GetHeaderBlock()->GetLoopMergeInst();
+  assert(loop_merge_inst &&
+         "Loop merge instruction could not be found after entering unroller "
+         "(should have exited before this)");
+
+  uint32_t loop_control_dont_unroll = 2;
+  uint32_t loop_control_index = 2;
+  loop_merge_inst->SetInOperand(loop_control_index, {loop_control_dont_unroll});
 }
 
-// Duplicate the |loop| body |factor| number of times while keeping the loop
-// backedge intact.
+// Duplicate the |loop| body |factor| - 1 number of times while keeping the loop
+// backedge intact. This will leave the loop with |factor| number of bodies
+// after accounting for the initial body.
 void LoopUnrollerUtilsImpl::Unroll(ir::Loop* loop, size_t factor) {
+  // If we unroll a loop partially it may not be safe to unroll it further.
+  MarkLoopControlAsDontUnroll(loop);
```

Comment:
Under which conditions?  This seems to always mark the loop unsafe to unroll again.

Reply:
This comes back to the current induction variable mechanism, one of the constraints we enforce is that the loop is bound by one induction variable which is modified by a single constant in the step instruction. We use the values from that to work out the number of iterations in the loop. Since the loop now has an induction variable with multiple step instructions it won't work and will incorrectly calculate the number of iterations.

Reply:
I can't believe I missed this.  That case 

```
#version 430

layout(location = 0) out float o;

void main(void) {
  float s = 0.0;
  for (int j = 0; j < 3; j += 1) {
    s += 1.0;
                j += 1;
  }
  o = s;
}
```

Will not unrolled correctly.

Reply:
No it shouldn't be getting unrolled at all. I'll make sure to check it doesn't and add a test if we don't already have one. It's one of the things we can support fairly easily with the more advanced induction variable detection code mentioned above